### PR TITLE
(Re)Enable NativeVarargs for CoreCLR

### PR DESCRIPTION
**Pull request by jashook**
This will allow coreclr run programs which have native varargs.
It also re-enables the ArgIterator type for managed to managed native
vararg calls which will allow the use of the __arglist keyword.

This change also deletes a test which expects an invalid program exception
for native varargs usage.

**Limitations:**

NYI statements have been added for all Unix Targets.

With this change __arglist (native varargs) will be supported, **but not
yet tested** on:

```
Amd64 Windows
x86 Windows
Arm32 Windows
Arm64 Windows
```

This change does not address re-enablabling native vararg tests. This will be done
in a seperate change.

See #18377 for on initial thoughts on varargs bringup on 
unix platforms.

### CODE REVIEW

File: src/jit/lclvars.cpp [hunk 1]
```diff
@@ -598,6 +598,18 @@ void Compiler::lvaInitUserArgs(InitVarDscInfo* varDscInfo)
                 isHfaArg = varTypeIsFloating(hfaType);
             }
         }
+        else if (info.compIsVarArgs)
+        {
+#ifdef _TARGET_UNIX_
+            // Currently native varargs is not implemented on non windows targets.
+            //
+            // Note that some targets like Arm64 Unix should not need much work as
```
Comment (janvorli) [inline src/jit/lclvars.cpp]: The ABI is not the same for ARM64 Unix - we've discussed it together some time ago. And with Amd64, things are even more complex on unix - the variable args can be passed in registers, the va_list is a special structure containing the indices of the gen purpose / fp registers, etc.
Am I missing something?

File: src/vm/callingconvention.h [hunk 2]
```diff
@@ -417,11 +417,10 @@ class ArgIteratorTemplate : public ARGITERATOR_BASE
 
 #ifdef _TARGET_AMD64_
 #ifdef UNIX_AMD64_ABI
-        PORTABILITY_ASSERT("ArgIteratorTemplate::IsVarArgPassedByRef");                
         return FALSE;
-#else // UNIX_AMD64_ABI
+#else // !UNIX_AMD64_ABI
```
Comment (janvorli) [inline src/vm/callingconvention.h]: The comment was correct - the convention in the VM is to always use the condition from the #ifdef, not the negation.

File: src/System.Private.CoreLib/src/System/ArgIterator.cs [hunk 3]
```diff
@@ -132,54 +131,5 @@ public override bool Equals(Object o)
         {
             throw new NotSupportedException(SR.NotSupported_NYI);
         }
-#else
-        public ArgIterator(RuntimeArgumentHandle arglist)
-        {
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/ArgIterator.cs]: You should keep this throwing implementation for Unix until/if the varargs are actually implemented for Unix

File: src/jit/lclvars.cpp (hunk 1 above)
Comment (jashook) [reply src/jit/lclvars.cpp]: The comment is addressing code generation. For arm64 unix the ABI for varargs is not different from the regular calling convention.

File: src/vm/callingconvention.h (hunk 2 above)
Comment (jashook) [reply src/vm/callingconvention.h]: Ok will address.

File: src/vm/jitinterface.cpp [hunk 4]
```diff
@@ -489,13 +489,6 @@ CEEInfo::ConvToJitSig(
         IfFailThrow(sig.GetCallingConvInfo(&data));
         sigRet->callConv = (CorInfoCallConv) data;
 
-        if ((isCallConv(sigRet->callConv, IMAGE_CEE_CS_CALLCONV_VARARG)) ||
```
Comment (jkotas) [inline src/vm/jitinterface.cpp]: We should continue throwing here for unix

File: src/System.Private.CoreLib/src/System/ArgIterator.cs (hunk 3 above)
Comment (jashook) [reply src/System.Private.CoreLib/src/System/ArgIterator.cs]: Is there a platform specific define that can be used here?

File: src/jit/lclvars.cpp (hunk 1 above)
Comment (janvorli) [reply src/jit/lclvars.cpp]: Ah, makes sense, I was somehow mislead into thinking you were comparing the actual windows and Unix abis.

File: src/System.Private.CoreLib/src/System/ArgIterator.cs (hunk 3 above)
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/ArgIterator.cs]: `PLATFORM_UNIX` or `PLATFORM_WINDOWS`

File: tests/src/baseservices/varargs/varargsupport.il [hunk 5]
```diff
@@ -1,114 +0,0 @@
-// Licensed to the .NET Foundation under one or more agreements.
```
Comment (jkotas) [inline tests/src/baseservices/varargs/varargsupport.il]: Should we keep these tests for Unix?

Comment (jashook) [reply tests/src/baseservices/varargs/varargsupport.il]: Personally I do not see the value of the test and conditionally excluding it by platform will be more work to undo later. If there are strong opinions to keep them then I can add them back.

File: src/vm/jitinterface.cpp [hunk 6]
```diff
@@ -489,12 +489,14 @@ CEEInfo::ConvToJitSig(
         IfFailThrow(sig.GetCallingConvInfo(&data));
         sigRet->callConv = (CorInfoCallConv) data;
 
+#ifdef _TARGET_UNIX_
```
Comment (jkotas) [inline src/vm/jitinterface.cpp]: `_TARGET_UNIX_` is JIT-specific define. Outside of JIT, we have `PLATFORM_UNIX`.

File: tests/src/baseservices/varargs/varargsupport.il (hunk 5 above)
Comment (jkotas) [reply tests/src/baseservices/varargs/varargsupport.il]: This test would have caught the wrong ifdef that I have commented on...

Are we going to implement varargs support on Unix for next release? If yes, I do not have a problem with deleting these tests now. If not, we should keep them.

Comment (jashook) [reply tests/src/baseservices/varargs/varargsupport.il]: Fair point will add them back thank you.

File: src/vm/jitinterface.cpp (hunk 6 above)
Comment (jashook) [reply src/vm/jitinterface.cpp]: Thank you for catching this.

File: src/jit/morph.cpp [hunk 7]
```diff
@@ -2737,9 +2737,18 @@ GenTreeCall* Compiler::fgMorphArgs(GenTreeCall* call)
...
-    bool callIsVararg = call->IsVarargs();
+#ifdef _TARGET_UNIX_
+    if (callIsVararg)
+    {
+        // Currently native varargs is not implemented on non windows targets.
+        //
+        // Note that some targets like Arm64 Unix should not need much work as
+        // the ABI is the same. While other targets may only need small changes
+        // such as amd64 Unix, which just expects RAX to pass numFPArguments.
+        NYI("Morhing Vararg call not yet implemented on non Windows targets.");
```
Comment (briansull) [inline src/jit/morph.cpp]: Spelling: Morhing => Morphing

File: src/jit/morph.cpp [hunk 8]
```diff
@@ -2737,9 +2737,18 @@ GenTreeCall* Compiler::fgMorphArgs(GenTreeCall* call)
...
+    }
 #endif
```
Comment (briansull) [inline src/jit/morph.cpp]: // _TARGET_UNIX_

**Pull request merged**